src/openms/source/QC/FragmentMassError.cpp

```diff
@@ -0,0 +1,291 @@
+      }
+
+
+      //---------------------------------------------------------------------
+      // CREATE THEORETICAL SPECTRUM
+      //---------------------------------------------------------------------
+
+      // theoretical peak spectrum
+      PeakSpectrum theo_spectrum;
+
+      // initialize a TheoreticalSpectrumGenerator
+      TheoreticalSpectrumGenerator theo_gen;
+
+      // get current parameters (default)
+      // default with b and y ions
+      Param theo_gen_settings = theo_gen.getParameters();
+
+
+      if (exp_spectrum.getPrecursors().empty() || exp_spectrum.getPrecursors()[0].getActivationMethods().empty())
+      {
+        throw Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No fragmentation method given.");
+      }
+
+      Precursor::ActivationMethod fm = (*exp_spectrum.getPrecursors()[0].getActivationMethods().begin());
+
+      if (fm == Precursor::ActivationMethod::ECD || fm == Precursor::ActivationMethod::ETD)
+      {
+        theo_gen_settings.setValue("add_c_ions", "true");
+        theo_gen_settings.setValue("add_z_ions", "true");
+        theo_gen_settings.setValue("add_b_ions", "false");
+        theo_gen_settings.setValue("add_y_ions", "false");
+      }
+
+      else if (!(fm == Precursor::ActivationMethod::CID || fm == Precursor::ActivationMethod::HCID))
+      {
+        throw Exception::InvalidParameter(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Fragmentation method is not supported.");
+      }
+
+      // set changed parameters
+      theo_gen.setParameters(theo_gen_settings);
+
+      // generate a-, b- and y-ion spectrum of peptide seq with charge
+      theo_gen.getSpectrum(theo_spectrum, seq, charge, charge);
+
+      //-----------------------------------------------------------------------
+      // COMPARE THEORETICAL AND EXPERIMENTAL SPECTRUM
+      //-----------------------------------------------------------------------
+
+      if (exp_spectrum.empty() || theo_spectrum.empty())
+      {
+        LOG_WARN << "The spectrum with " + String(exp_spectrum.getRT()) + " is empty." << "\n";
+        return;
+      }
+
+      auto exp_spectrum_filtered(exp_spectrum);
+      window_mower_filter.filterPeakSpectrum(exp_spectrum_filtered);
+
+      // stores ppms for one spectrum
+      DoubleList ppms{};
+
+      // exp_peak matching to previous theo_peak
+      double current_exp = std::numeric_limits<double>::max();
+      
+      // minimal ppm
+      double ppm = std::numeric_limits<double>::max();
+
+      for (const Peak1D& peak : theo_spectrum)
+      {
+        const double theo_mz = peak.getMZ();
+        Size index = exp_spectrum_filtered.findNearest(theo_mz);
+        const double exp_mz = exp_spectrum_filtered[index].getMZ();
+        
+        const double mz_tolerance = (tolerance_unit=="ppm") ?  Math::ppmToMass(tolerance, theo_mz) : tolerance;
```

Comment:
see why bool is better? you check for `== "ppm"`, but you pass `mz` (in test) and `Da` (in TOPP) ... dangerous